SDKS-3938: Migrate device client

ancheetah opened the pull request:
# JIRA Ticket

https://pingidentity.atlassian.net/browse/SDKS-3938

## Description

- Copied device client package from legacy sdk to unified sdk
- adds manual e2e test scripts

No changeset. Added device client package and e2e app to changeset ignore

cerebrl commented on packages/device-client/src/lib/types/oath.types.ts:
```diff
@@ -5,11 +5,13 @@
 export type OathDevice = {
-  id: string;
+  _id: string;
+  deviceManagementStatus: boolean;
   deviceName: string;
   uuid: string;
-  createdDate: Date;
-  lastAccessDate: Date;
+  createdDate: number;
+  lastAccessDate: number;
+  _rev: string;
```
This reminds me of a MongoDB document. I wonder if they are using a NoSQL data store for storying this info.

cerebrl commented on e2e/device-client-app/src/autoscript.ts:
```diff
@@ -0,0 +1,129 @@
+} from '@forgerock/javascript-sdk';
+import { deviceClient } from '@forgerock/device-client';
+import { delay as rxDelay, mergeMap } from 'rxjs/operators';
+import { from } from 'rxjs';
```
Since this is a new repo, I think it's worth asking if the use of RxJS is needed. We never really utilized it in the old repo in any way that provided enough value to offset its cost.

What do you all think? Should we just toss RxJS and move to async-await?

ancheetah replied:
From what I can tell it looks like rxjs was only used for the autoscripts in the old repo. So I guess the question is also do we plan to use rxjs for any other autoscript tests that we plan to port over?

I don't have a strong opinion but I do think rxjs improves readability by grouping together different steps in the test flow. However, that might not be a good enough reason to install an additional dependency.

cerebrl replied:
> [...] I do think rxjs improves readability by grouping together different steps in the test flow.

Okay, that's reasonable. Let's see what @ryanbas21 has to say. I personally don't mind RxJS, but I just want to make sure we critically consider it, rather than just copy it over.

cerebrl replied:
I'm fine with transitioning over to Effect. I was actually thinking about that, and wondered if you'd mention it. We are already using it, like you said, and will likely introduce more of it in our code. So, this would be a good place to start. 

@ryanbas21 Would you be able to give AJ a simple code translation of what she has to Effect?

@ancheetah I'll give you the last right of refusal. This is just an e2e test, so I don't want to overly stress how it's implemented. If you

cerebrl reviewed (commented):
I think this is really well done. I like the reuse of the autoscript for this particular context. It's super clean. Let's just discuss whether we want to carryover RxJS into this repo or not.

cerebrl commented on packages/device-client/src/lib/mock-data/msw-mock-data.ts:
Ah, got it. Thanks. I'd rather avoid naming files with the underlying tech we use. Plus, we have a convention for naming mock files, so let's do that instead. For example: if the test is named `device.store.test.ts`, then the corresponding mock data would be named `device.store.mock.ts`.

cerebrl commented on packages/device-client/src/lib/device.store.ts:
```diff
@@ -43,16 +53,20 @@ export const deviceClient = (config: ConfigOptions) => {
-        return response.data;
+          return response.data.result;
+        } catch (error) {
+          return { error };
```
@ryanbas21 Should this throw the `GenericError` type that we settled on in DaVinci Client? My thought is that we should probably be consistent with errors throughout this SDK suite.

ancheetah commented:
- updated device client test user to avoid conflicts with Stoyan's automated tests
- moved msw handlers to their own file
- updated device client delete methods to return null
- updated device client integration tests
- removed rxjs from e2e tests and replaced with Effect

ancheetah commented on packages/device-client/src/lib/device.store.ts:
```diff
@@ -183,59 +225,140 @@ export const deviceClient = (config: ConfigOptions) => {
-          return undefined;
-        }
+          if (error) {
+            handleError(error, 'Failed to delete device: ');
```
I was just trying not to fix what wasn't broken. The other get/update methods originally didn't handle errors from RTK. They just check for data and if it's missing it'll throw a 'response did not contain data' error. To be more consistent I can update the other methods to check for errors first before returning data.

cerebrl replied:
Fair enough. We can improve this in the future.

cerebrl reviewed (approved):
I think this is good for the migration story. I'll make a few improvement stories in Jira to continue iterating on this foundation.
